### ISSUE TITLE
Major performance regression in certain files starting with 1.9.4 -- 100% CPU for about 1 minute

### ISSUE THREAD

**Issue by sean-mcmanus** [jeff-hykin/cpp-textmate-grammar#218]
Certain .cpp files make VS Code unusuably slow due to high 100% CPU usage on one core (extension host process) starting with 1.9.4 -- the issue doesn't repro with 1.9.3. It repros without our C/C++ extension. Could it be caused by https://github.com/jeff-hykin/cpp-textmate-grammar/pull/199/files ?

The repro we have has 2349 lines in it. Let me know if you guys need help getting a better repro.

I'm using the latest VS Code Insiders.

**Comment by sean-mcmanus** [comment]
I ran the profiler -- it's showing vscode-textmate matchRule taking all the time.

**Comment by sean-mcmanus** [comment]
I got the repro down to 91 lines. Invoking Developer: Inspect TM Scope causes major blockage "forever" until I reload.

I don't see anything unusual in the code.

**Comment by matter123** [comment]
The reproduction case would certainly be helpful. The only significant difference between v1.9.3 and v1.9.4 is support for default arguments in function definitions https://github.com/jeff-hykin/cpp-textmate-grammar/commit/a048985cefe580ac5ba6c26a202ed69cacdfe714, and all those patterns were preexisting.

**Comment by sean-mcmanus** [comment]
Here's a repro -- I've replaced all the identifiers with "a". Freeze is 10-15 seconds -- more with Developer: Inspect TM Scope enabled.

```


static void a(
                           a    a,
                           a a,
                           a a,
                           a a,
                           a a,
                           bool                       a,
                           bool                       a,
                           bool                       a,
                           bool                       a)

{
  a a;


  a(a, a,
                                 a,
                                 a,
                                 a,
                                 a,
                                 a,
                                 a,
                                 a);
  a->a(&a);
  for (auto a : a) {
    auto &a = a.a;
    auto a = a.a;

    if (a && a) {
      switch (a) {
        case a:
          a->a(a->a());
          break;
        case a:
          a->a(a->a());
          break;
        case a:
          a->a(a->a());
          break;
        case a:
          a->a(a->a());
          break;
      }  
    }
    a(a, a,
                                   a,
                                   a,
                                   a,
                                   a,
                                   a,
                                   a,
                                   a);
  } 
}  

static void a(
                      a                    a,
                      a           a,
                      a a,
                      a                                 a)

{
  a a;
  a a =0;
                               a->a()->a();

  a(a->a());
  a->a(a);
  a->a(a);
  a(a, a, a,
                                 a,
                                 a);
  a->a(&a);
  for (auto a: a) {
    a(a->a(), a, a,
                                                                   a);
  }
}

```

**Comment by matter123** [comment]
@jeff-hykin The issue seems to be that function_definition is allowed inside of the function parameters. vscode-textmate spends about half of its time (4700ms) to determine that `:function_definition` is not a match. Do you know why its there?

**Comment by matter123** [comment]
It looks like the performance issue is with generateScopeResolutionFinder. when the pattern is forced to be atomic the total time becomes .91s. This causes other issues with matching so it's not a solution, but hopefully, the scope resolution generator can be made non-backtracking.

**Comment by jeff-hykin** [comment]
> vscode-textmate spends about half of its time (4700ms) to determine that :function_definition is not a match. Do you know why its there?

No I don't, and I can't think of a reason why it would intentionally be there. If removing it solves the issue, then we should definitely remove it.

It's still unclear to me why :function_definition would cause this, especially with backtracking disabled. It could be related to the look-arounds. I'm going to be without a computer tomorrow so I'm going to try to look into the issue now.

**Comment by jeff-hykin** [comment]
Thanks @sean-mcmanus for posting the example code

**Comment by jeff-hykin** [comment]
@matter123 

This is strange 
This code is not a problem:
```
static void a(
    bool a){
  a(a, a, a, a, a);
}
```
However this is near unusably slow:
```
static void a(
    bool                                                                     a){
  a(a, a, a, a, a);
}
```

Something is going on with the matching of whitespace.

**Comment by jeff-hykin** [comment]
Fix pushed with version `v1.11.2`
The fix corrects the `:function_call_context` and limits it's internal patterns to be much closer to what they should be. The performance issue seems to have disappeared. 

With that said, @matter123 I'm still not sure what was causing it to be so slow by simply adding whitespace. On another day I'm going to try to find the cause.

**Comment by sean-mcmanus** [comment]
Thanks, that fixed the issue with my 2.4k line file as well.
